World Info chat-independent data matching

BismuthGlass opened the pull request:
Sometimes it is useful to match World Info entries across character data such as description, personality and so on. This allows for dictionary style Lorebooks with multiple definitions which can be switched on or off according to the character being used. It may also occasionally be useful to match WI Entries on data that is not seen by the LLM, or even on the current user persona.

This feature adds extra World Info Entry options which allow for matching on the following chat-independent data:

- User persona
- Character description
- Character personality
- Character depth prompt (Character's Notes)
- Scenario
- Creator notes

The feature works by adding each of the fields to the World Info scan string, depending on which fields are active. This means that these matches use the exact same mechanism as regular chat matching, which means other options such as trigger chance, cooldown, etc. should work just fine.

To accommodate these new toggles, I have grouped them together under a new **Additional Matching Sources** dropdown:

![image](https://github.com/user-attachments/assets/06ee69ef-b294-4062-bc98-d25b502d1024)

Let me know what you think.

## Arguments against

> Add individual character names / persona name to entries

This doesn't work well in certain cases because every time we create a new character that is supposed to use the lorebook, we have to edit the lorebook to include this character. This is tedious, time consuming, and polutes the Lorebook.

> Use tags instead

Tags should, in my opinion, be used for organization purposes. Having a huge set of tags in one character can hurt character list maintainability, and hinder glance value. With smaller Lorebooks tags could be an appropriate solution, but when you are looking to have 50+ entries it doesn't seem very viable.

## Checklist:

- [x] I have read the [Contributing guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).

BismuthGlass committed:
Add controls for different WI matching targets

To accomodate new settings, the WI Entry panels were reworked
slightly to add a drawer. Both Global and Entry settings are
present.

BismuthGlass committed:
Add WIEntry support for new match options

BismuthGlass committed:
Change matchCharacterMetadata to matchCreatorsNotes

BismuthGlass committed:
Fix field names

BismuthGlass committed:
Change matchCreatorNotes name

BismuthGlass committed:
Add WIGlobalScanData type

BismuthGlass committed:
Change match options to checkboxes

BismuthGlass committed:
Revert global WI settings changes

BismuthGlass committed:
Implement WI matching on global data

BismuthGlass committed:
Change character note to character depth prompt

BismuthGlass committed:
Pass global scan data to WI prompt generator

BismuthGlass committed:
Remove newlines

BismuthGlass committed:
Fix save bug

BismuthGlass committed:
Fix world info entry saves for match fields

BismuthGlass committed:
Fix matching issues for depth prompt

BismuthGlass committed:
Change double quotes to single quotes

Cohee1207 commented:
I'm almost sure this either has to be WI book-specific or character-specific setting... Being global means you're likely hit lots of false positives in some areas.

BismuthGlass commented:
@Cohee1207 At the moment it is WI book entry specific. I actually was thinking of implementing global toggles for all of those as the ones for **Include Names**, **Case-sensitive** and so on, but ultimately decided against it. I also pondered making it a character specific setting, by adding "include in world info scanning" toggles for each field, but I felt that would limit the power of the feature for little gain. Unless that's not what you meant by it?

As a quick example of how this could be used (how I am currently using it), I have a bunch of personality traits defined which I can match against particular strings in the **Personality** entry. This allows me to really easily re-use these definitions for all my characters by simply having the Lorebook always active, and adding a few keywords, as opposed to fiddling with tags:

```
ATTRIBUTES: #shy #sarcastic #nature-lover(deserts)
```

I haven't used the other match modes yet, but I added them for the sake of completion.

Cohee1207 commented:
The idea is good, but I don't agree with the UI choices. "Activation settings" is a confusing name since every field except content is directly related to entry activation (content too if recursive is enabled but I digress). What if you group these new checkboxes under a separate drawer ("Additional Matching Sources")? That would not break muscle memory of users and still utilize a nested drawers concept. See below. Ignore that half of the UI is missing, it's just a mockup.

<img width="745" alt="image" src="https://github.com/user-attachments/assets/c29b8985-4b4f-4524-90d8-8559cd07e3c2" />

Cohee1207 commented on public/scripts/world-info.js:
```diff
@@ -2191,6 +2235,12 @@ export const originalWIDataKeyMap = {
     'matchWholeWords': 'extensions.match_whole_words',
     'useGroupScoring': 'extensions.use_group_scoring',
     'caseSensitive': 'extensions.case_sensitive',
+    'matchPersonaDescription': 'extensions.match_persona_description',
```
It's not at all obvious, but these fields need to be written on backend, otherwise they won't be sharable with card-embedded lorebooks.

https://github.com/SillyTavern/SillyTavern/blob/689637b36c178478545b46a4bec0f25ae3b97471/src/endpoints/characters.js#L662

BismuthGlass committed:
Revert original settings to their former place

BismuthGlass committed:
Fix character card lorebook imports / exports

BismuthGlass committed:
Revert world_entry_form_control css class changes

BismuthGlass committed:
Change handleOptionalSelect name

BismuthGlass committed:
Add missing machCharacterPersonality to World Info Definition

BismuthGlass committed:
Add back chat param

BismuthGlass committed:
Move matching sources to bottom of entry editor

BismuthGlass commented:
@Cohee1207  thanks for the feedback!

I have made a few changes and a little bit of cleanup. I have added the matching controls to the bottom of the editor under a dropdown as you suggested, making sure to revert the changes to the other components, of course (updated the picture in the description to reflect this too!) I have also gone through the character lorebook imports and it should now be working correctly. Let me know what you think.

Cohee1207 commented on public/index.html:
```diff
@@ -6291,6 +6291,52 @@ <h4>
                             </div>
+                            <div class="inline-drawer wide100p flexFlowColumn">
+                                <div class="inline-drawer-toggle inline-drawer-header userSettingsInnerExpandable">
+                                    <strong>Additional Matching Sources</strong>
```
Please make the text localizable

Cohee1207 commented on public/script.js:
```diff
@@ -3162,6 +3164,7 @@ export function getCharacterCardFields({ chid = null } = {}) {
     result.jailbreak = power_user.prefer_character_jailbreak ? baseChatReplace(character.data?.post_history_instructions?.trim(), name1, name2) : '';
     result.version = character.data?.character_version ?? '';
     result.charDepthPrompt = baseChatReplace(character.data?.extensions?.depth_prompt?.prompt?.trim(), name1, name2);
+    result.creatorNotes = baseChatReplace(character.data?.creator_notes?.trim(), name1, name2);
```
Does it make sense to add it as a macro? Since the rest of the fields are valid macros as well.

Cohee1207 commented on public/script.js:
```diff
@@ -4143,7 +4147,15 @@ export async function Generate(type, { automatic_trigger, force_name2, quiet_pro
+        scenario: scenario,
+        creatorNotes: creatorNotes,
+    };
+    const { worldInfoString, worldInfoBefore, worldInfoAfter, worldInfoExamples, worldInfoDepth } = await getWorldInfoPrompt(globalScanData, chatForWI, this_max_context, dryRun);
```
Please don't add new params at the beginning of a functions. This breaks backwards compat with external exports

Cohee1207 commented on public/scripts/world-info.js:
```diff
@@ -165,10 +187,12 @@ class WorldInfoBuffer {
 
     /**
      * Initialize the buffer with the given messages.
+     * @param {WIGlobalScanData} globalScanData Chat independent context to be scanned
```
Same here. Try adding new params at the end

Cohee1207 commented on public/scripts/world-info.js:
```diff
@@ -4848,6 +4927,12 @@ export function convertCharacterBook(characterBook) {
             sticky: entry.extensions?.sticky ?? null,
             cooldown: entry.extensions?.cooldown ?? null,
             delay: entry.extensions?.delay ?? null,
+            matchPersonaDescription: entry.extensions?.matchPersonaDescription ?? false,
```
Extension keys don't match the actual recorded values.

Cohee1207 replied:
(should be entry.extensions.match_persona_description)

Cohee1207 commented on public/scripts/world-info.js:
```diff
@@ -2191,6 +2235,12 @@ export const originalWIDataKeyMap = {
     'useGroupScoring': 'extensions.use_group_scoring',
     'caseSensitive': 'extensions.case_sensitive',
+    'matchPersonaDescription': 'extensions.match_persona_description',
+    'matchCharacterDescription': 'extensions.match_character_description',
```
Extension keys don't match the actual recorded values.

BismuthGlass committed:
Add i18n for Additional Matching Sources header

BismuthGlass committed:
Move new globalScanData args to the end

BismuthGlass committed:
Add creatorNotes macro expansion

BismuthGlass committed:
Add type data for v2DataWorldInfoEntryExtensionInfos match fields

BismuthGlass committed:
Change field names to match required fields

BismuthGlass committed:
Add pointer cursor to Additional Matching Sources drawer header

BismuthGlass committed:
Add extra field docs

BismuthGlass commented:
@Cohee1207 By macro, I assume you mean a {{creatorNotes}} macro? It wouldn't be that useful for me, but I added it.

- Moved new parameters to the end
- Added i18n for the Additional Matching Sources header
- Added pointer cursor to the Additional Matching Sources header
- Renamed match_* field keys to align with the rest of the fields
- Added the new fields to the `v2DataWorldInfoEntryExtensionInfos` type
- Added `creatorNotes` to story string params, allowing for `{{creatorNotes}}` expansion

Cohee1207 reviewed (approved):
The code looks ok now. I'll verify the actual functionality as soon as I can.

Cohee1207 committed:
Move creatorNotes macro init

Cohee1207 committed:
Do not append empty joiners

Wolfsblvt commented:
Not sure about the checkboxes all below each other (lot's of wasted space on the right), but as the area is collapsed by default this should still be okay.

Feature looks nice, idea is good and was requested often. For the same reasons you noted why in some scenarios char filter or tags are cumbersome.

I want to give this a spin to test too.

Cohee1207 commented:
Tested, looks good from my side, fixed a couple of trivial things. Waiting for Wolf's testing results.

BismuthGlass commented:
@Wolfsblvt I could also split the options into two columns, or we could keep it like that and then perhaps use the extra space for new features in the future.

Cohee1207 commented:
Two columns would look nicer, but likely requires to remove "Match" from every label ("Matching Sources" should provide enough context).

Wolfsblvt commented:
If you remove "Match" from each of them, it could even become toggle-able buttons with labels. Like tags.
Could check the recent `/buttons` implementation for the multiple toggleable buttons. Just in a single line, not one after the other.

Not 100% sure if that's better than checkboxes, but we could save a lot of space by removing "Match" and the checkbox if the label itself is the checkbox.

Cohee1207 committed:
Align matching sources in two columns

Cohee1207 commented:
Grouped into two columns. Checkbox names trimmed. Good to merge now..

Cohee1207 commented:
Oh yeah, forgot to mention that someone has to write the docs for this one.

https://github.com/SillyTavern/SillyTavern-Docs/blob/main/Usage/worldinfo.md
